### PR TITLE
Prevent duplicate channel cache entries when prepending from query

### LINKED ISSUE #3475: prependChanges can result in multiple revisions for a doc in channel cache
When issuing a view query to backfill the channel cache, any results are prepended to the channel cache when the channel cache has room available.

However, the channel cache is not locked between the original query that triggered the view query, and the subsequent prepend.  This leaves the possibility of the following:

1. Document foo exists w/ rev 1-abc in channel A
2. User issues changes request for channel A, triggering view backfill
3. SG issues stale=false channels query
4. Document foo is updated to rev 2-def (still in channel A)
5. Doc foo 2-def arrives over DCP feed, is added to channel cache for channel A
6. View query completes, includes foo rev 1-abc (stale=false isn't guaranteed to see changes made after query is issued)
7. View results are prepended to cache, so cache contains both 1-abc and 2-def for channel A.

This won't cause any correctness issues, but can result in additional replication overhead, and sub-optimal cache utilization (other docs could be ejected from the cache in favour of the duplicates).

### PR DESCRIPTION
**Pull request by adamcfraser**
When prepending entries to the channel cache, enforce that prepended docs don't have a later revision already present in the cache.  Can occur when new cache entries are added while the query used to backfill/prepend the cache is running.

Fixes #3475 

Integration tests:
http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/624/
http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/625/

### CODE REVIEW

**Comment by adamcfraser** [comment]
TestChannelView failed in a way I haven't seen before in the xattr integration test - pushed some additional diagnostics and rerunning:
http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/641

**Comment by tleyden** [comment]
>  pushed some additional diagnostics and rerunning:

Did you manage to get to the bottom of it?  Or was it just a sporadic failure?

**Comment by adamcfraser** [comment]
Looks like a sporadic failure.  Reran 5x here without failure:
http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/646/

File: db/channel_cache.go [hunk 1]
```diff
@@ -381,27 +380,48 @@ func (c *channelCache) prependChanges(changes LogEntries, changesValidFrom uint6
...
+					}
+
+					numToPrepend := len(entriesToPrepend)
+					if numToPrepend > 0 {
+						c.logs = append(entriesToPrepend, c.logs...)
+						base.Infof(base.KeyCache, "  Added %d entries from query (#%d--#%d) to cache of %q",
+							numToPrepend, entriesToPrepend[0].Sequence, entriesToPrepend[numToPrepend-1].Sequence, base.UD(c.channelName))
 					}
-					base.Debugf(base.KeyCache, " Backfill cache from view c.validFrom from %v -> %v",
+					base.Debugf(base.KeyCache, " Backfill cache from query c.validFrom from %v -> %v",
```
Comment (tleyden) [inline db/channel_cache.go:421]: I think the wording could be a bit clearer .. maybe: "Completed channel cache backfill from query.  Update cache validFrom sequence from %v -> %v"

**Review by tleyden — APPROVED** [review]
LGTM.  One suggestion on improving wording in logs.

**Pull request merged**